### PR TITLE
SONARJAVA-4678: mplement S6831: @Qualifier should not be used on @Bean methods

### PR DESCRIPTION
**Pull request by ADarko22**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/spring/AvoidQualifierOnBeanMethodsCheck.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    var methodTree = (MethodTree) tree;
+
+    var beanAnnotation = getAnnotation(methodTree, BEAN_ANNOTATION);
+    var qualifierAnnotation = getAnnotation(methodTree, QUALIFIER_ANNOTATION);
+
+    if(beanAnnotation != null && qualifierAnnotation != null) {
+      QuickFixHelper.newIssue(context)
+       .forRule(this)
+        .onTree(qualifierAnnotation)
+        .withMessage("Remove this redundant \"@Qualifier\" annotation")
```
Comment (dorian-burihabwa-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/spring/AvoidQualifierOnBeanMethodsCheck.java:56]: Let's add a `.` to finish the issue message
```suggestion
        .withMessage("Remove this redundant \"@Qualifier\" annotation.")
```

**Review by dorian-burihabwa-sonarsource — CHANGES_REQUESTED** [review]

File: java-checks/src/main/java/org/sonar/java/checks/spring/AvoidQualifierOnBeanMethodsCheck.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+      .orElse(null);
+  }
+
+  private static List<JavaQuickFix> getQuickFix(MethodTree methodTree, AnnotationTree qualifierAnnotation) {
+    List<JavaQuickFix> quickFixes = new LinkedList<>();
+
+    // quick fix only for @Qualifier annotations without arguments or with argument that matches the method name
+    if(isFixable(methodTree, qualifierAnnotation)) {
+      var quickFix = JavaQuickFix.newQuickFix("Remove \"@Qualifier\"")
+        .addTextEdit(JavaTextEdit.removeTree(qualifierAnnotation))
```
Comment (ADarko22) [inline java-checks/src/main/java/org/sonar/java/checks/spring/AvoidQualifierOnBeanMethodsCheck.java:80]: quick fix tested locally and works; anyway comments on the same line are not remove (i.e. the // Noncompliant int he sample)

File: java-checks/src/main/java/org/sonar/java/checks/spring/AvoidQualifierOnBeanMethodsCheck.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    var qualifierAnnotationValue = "";
+
+    if(argument.is(Tree.Kind.ASSIGNMENT)) {
+      qualifierAnnotationValue = ((LiteralTreeImpl) ((AssignmentExpressionTreeImpl) argument).expression()).value().replace("\"", "");
+    }
+    else if(argument.is(Tree.Kind.STRING_LITERAL)) {
+      qualifierAnnotationValue = ((LiteralTreeImpl) argument).token().text().replace("\"", "");
+    }
+
+    return qualifierAnnotationValue;
```
Comment (dorian-burihabwa-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/spring/AvoidQualifierOnBeanMethodsCheck.java:113]: I understand checking the types twice just to be sure but in practice I am not sure if we are likely to see other types of arguments. Have you encountered any other while testing?

File: java-checks/src/main/java/org/sonar/java/checks/spring/AvoidQualifierOnBeanMethodsCheck.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    var methodName = methodTree.simpleName().name();
+    return getQualifierAnnotationValue(arguments).equals(methodName);
+  }
+
+  private static String getQualifierAnnotationValue(Arguments arguments) {
+    var argument = arguments.get(0);
+    var qualifierAnnotationValue = "";
+
+    if(argument.is(Tree.Kind.ASSIGNMENT)) {
+      qualifierAnnotationValue = ((LiteralTreeImpl) ((AssignmentExpressionTreeImpl) argument).expression()).value().replace("\"", "");
```
Comment (dorian-burihabwa-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/spring/AvoidQualifierOnBeanMethodsCheck.java:107]: Is it possible to set the value of the assignment expression using a constant?
```java
class A {
  static final String CONSTANT = "foo";
  @Bean  
  @Qualifier(value = CONSTANT)
  public String foo() { return "Hello"; }
}
```
I know that it is possible for the `@Variable` annotation. We should just make sure it is safe for our code to assume that we always have a literal

**Review by dorian-burihabwa-sonarsource — CHANGES_REQUESTED** [review]
Looking nice on the QF front, let's make sure we cover all our angles and fix the open issues

**Review by dorian-burihabwa-sonarsource — APPROVED** [review]
LGTM 👍🏿 Thanks for the changes

**Pull request merged**